leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/CaliforniumLwM2mRequestSender.java

```diff
@@ -47,6 +51,7 @@
     private final Set<Endpoint> endpoints;
     private final ObservationRegistry observationRegistry;
     private final LwM2mModelProvider modelProvider;
+    private final Map<String/* registration id */, ArrayList<Request> /* pending coap Request */> pendingRequests = new HashMap<>();
```

Comment:
Totally agree, we should  keep focus on this bugfix PR. 
(About the use of Client object, I see what you mean, but this will not work with a not in memory implementation because you could have 2 `Client` instances which will represent the same Client/Registration. Each get on ClientRegistry will return a new instance)